app/ui/src/app/integration/integration_detail/integration_activity/integration-activity.component.ts

```diff
@@ -32,46 +33,44 @@ export class IntegrationActivityComponent implements OnInit {
     this.fetchActivities();
   }
 
-  stepName(step: Step): string {
-    if (!step) {
-      return 'n/a';
-    }
-    if (step.name) {
-      return step.name;
-    }
-    if (step.action && step.action.name) {
-      return step.action.name;
-    }
-    return 'n/a';
+  fetchStepName(step: Step): string {
+    return step && step.name ? step.name : step && step.action && step.action.name ? step.action.name : 'n/a';
   }
 
   fetchActivities(): void {
     this.onRefresh = true;
     this.onError = false;
-    this.integrationSupportService
-      .requestIntegrationActivity(this.integration.id)
-      .do(activitities => {
-        // TODO: In a real, efficient RFP environment, this should be performed as a
-        //       one step operation within the reducer data logic and never within a component
-        activitities.forEach(activity => {
-          if (activity.steps && Array.isArray(activity.steps)) {
-            activity.steps.forEach(step => {
-              // XXX: ANTIPATTERN AHEAD. The following code block mutates an object state
-              const integrationStep = this.integration.steps.find(_integrationStep => _integrationStep.id == step.id);
-              step.name = this.stepName(integrationStep);
-              step.isFailed = step.failure && step.failure.length > 0;
-              const errorMessages = [null, ...step.messages, step.failure].filter(messages => !!messages);
-              step.output = errorMessages.length > 0 ? errorMessages.join('\n') : null;
-            });
-          }
-        });
-
-        return activitities;
-      })
-      .subscribe(
-        activities => this.updateActivities(activities),
-        error => this.handleError(error)
-      );
+
+    const activities$ = this.integrationSupportService.requestIntegrationActivity(this.integration.id)
+    const integrationDeployments$ = this.integrationSupportService.getDeployments(this.integration.id); //.map(foo => foo['items']) as Observable<IntegrationDeployment[]>;
+
+    forkJoin<[Activity[], IntegrationDeployment[]]>([activities$, integrationDeployments$]).map(results => {
+      const activitities = results[0];
+      const integrationDeployments = results[1];
+
+      activitities.forEach(activity => {
+        if (activity.steps && Array.isArray(activity.steps)) {
+          activity.steps.forEach(step => {
+            const integrationStep = integrationDeployments
+              .find(deployment => deployment.version === +activity.ver)
+              .spec
+              .steps.find(integrationStep => integrationStep.id == step.id);
+            
+            step.name = this.fetchStepName(integrationStep);
+            step.isFailed = step.failure && step.failure.length > 0;
+            
+            const errorMessages = [null, ...step.messages, step.failure].filter(messages => !!messages);
```

Comment:
Please take a minute to review lines 44 and 45: In order to correlate data we need to fetch both the integration activity and all deployments available, refreshing data upon clicking on the refresh button.

Loading the step data upon bootstrapping components dynamically (which can only be done upon toggling the component +ngIf's structural directive) would force to fetch ALL the deployments everytime we render a steps table, and such data would be lost if the component is then hidden from the view.

We can wrap the steps data table into a standalone component, but it would become a dumb component either way (this is, a component that does not initiate data logic by itself), so the logic you see above would be still enforced in the parent component. This is, this one.

Reply:
Instead of fetching all deployments, you could fetch the data for a single deployment using the following API:

Method |Path 
--------|-----
GET | /api/v1/integrations/{id}/deployments/{version}

Reply:
But yeah the code seems like should correlate the right steps.